db/crud.go

```diff
@@ -1356,15 +1357,36 @@ func (db *DatabaseCollectionWithUser) PutExistingCurrentVersion(ctx context.Cont
 						return nil, nil, false, nil, addNewerVersionsErr
 					}
 				} else {
-					base.InfofCtx(ctx, base.KeyCRUD, "conflict detected between the two HLV's for doc %s, incoming version %s, local version %s", base.UD(doc.ID), newDocHLV.GetCurrentVersionString(), doc.HLV.GetCurrentVersionString())
-					// cancel rest of update, HLV needs to be sent back to client with merge versions populated
-					return nil, nil, false, nil, base.HTTPErrorf(http.StatusConflict, "Document revision conflict")
+					// temp remove history alignment here, pending CBG-4791
+					alignRevTrees = false
+					revTreeConflictChecked = true
+					previousRevTreeID = doc.CurrentRev
+					newGeneration = prevGeneration + 1
+					base.InfofCtx(ctx, base.KeyCRUD, "conflict detected between the two HLV's for doc %s, incoming version %v, local version %v", base.UD(doc.ID), newDocHLV.ExtractCurrentVersionFromHLV(), doc.HLV.ExtractCurrentVersionFromHLV())
+					if conflictResolver == nil {
+						// cancel rest of update, HLV is in conflict and no resolver is present
+						return nil, nil, false, nil, base.HTTPErrorf(http.StatusConflict, "Document revision conflict")
+					}
+					// resolve conflict
+					var newHLV *HybridLogicalVector
+					newHLV, err = db.ResolveHLVConflict(ctx, doc, newDoc, conflictResolver)
+					if err != nil {
+						base.InfofCtx(ctx, base.KeyCRUD, "Failed to resolve HLV conflict for doc %s, error: %v", base.UD(doc.ID), err)
+						return nil, nil, false, nil, err
+					}
+					conflictResolved = true // mark that we resolved the conflict so we don't overwrite the HLV MV below
+					// overwrite the existing HLV with the new one
+					doc.HLV = newHLV
 				}
 			}
 		}
-		// populate merge versions
-		if newDocHLV.MergeVersions != nil {
-			doc.HLV.MergeVersions = newDocHLV.MergeVersions
+		// we only want to populate the merge versions if we have not already resolved the conflict above given the
+		// special handling for conflict cases around merge versions.
+		if !conflictResolved {
+			// populate merge versions
+			if newDocHLV.MergeVersions != nil {
+				doc.HLV.MergeVersions = newDocHLV.MergeVersions
+			}
 		}
```

Comment:
I'm reasonably sure that this logic is just completely wrong, and I am going to move this into its own function as part of https://github.com/couchbase/sync_gateway/pull/7704. This problem pre-exists this PR.

I want to move the updating the HLV into its own function that can be used by blip tester and crud API.

If this code is ever called when a merge version is created, this is wrong, as you've noted.